src/core/net/srp_client.cpp

```diff
@@ -1552,112 +1535,207 @@ Error Client::AppendHostDescriptionInstruction(MsgInfo &aInfo)
     key.SetProtocol(Dns::KeyRecord::kProtocolDnsSec);
     key.SetAlgorithm(Dns::KeyRecord::kAlgorithmEcdsaP256Sha256);
     key.SetLength(sizeof(Dns::KeyRecord) - sizeof(Dns::ResourceRecord) + sizeof(Crypto::Ecdsa::P256::PublicKey));
-    SuccessOrExit(error = aInfo.mMessage->Append(key));
-    SuccessOrExit(error = aInfo.mKeyInfo.GetPublicKey(publicKey));
-    SuccessOrExit(error = aInfo.mMessage->Append(publicKey));
-    aInfo.mRecordCount++;
+    SuccessOrExit(error = mMessage->Append(key));
+    SuccessOrExit(error = mKeyInfo.GetPublicKey(publicKey));
+    SuccessOrExit(error = mMessage->Append(publicKey));
+    mRecordCount++;
 
 exit:
     return error;
 }
 
-Error Client::AppendDeleteAllRrsets(MsgInfo &aInfo) const
+Error Client::Update::AppendDeleteAllRrsets(void)
 {
     // "Delete all RRsets from a name" (RFC 2136 - 2.5.3)
     // Name should be already appended in the message.
 
+    Error               error;
     Dns::ResourceRecord rr;
 
     rr.Init(Dns::ResourceRecord::kTypeAny, Dns::ResourceRecord::kClassAny);
     rr.SetTtl(0);
     rr.SetLength(0);
 
-    return aInfo.mMessage->Append(rr);
+    SuccessOrExit(error = mMessage->Append(rr));
+    mRecordCount++;
+
+exit:
+    return error;
 }
 
-Error Client::AppendHostName(MsgInfo &aInfo, bool aDoNotCompress) const
+Error Client::Update::AppendHostName(bool aDoNotCompress)
 {
     Error error;
 
     if (aDoNotCompress)
     {
         // Uncompressed (canonical form) of host name is used for SIG(0)
         // calculation.
-        SuccessOrExit(error = Dns::Name::AppendMultipleLabels(mHostInfo.GetName(), *aInfo.mMessage));
-        error = Dns::Name::AppendName(mDomainName, *aInfo.mMessage);
+        SuccessOrExit(error = Dns::Name::AppendMultipleLabels(mHostName, *mMessage));
+        error = Dns::Name::AppendName(mDomainName, *mMessage);
         ExitNow();
     }
 
     // If host name was previously added in the message, add it
     // compressed as pointer to the previous one. Otherwise,
     // append it and remember the offset.
 
-    if (aInfo.mHostNameOffset != MsgInfo::kUnknownOffset)
+    if (mHostNameOffset != Update::kUnknownOffset)
     {
-        ExitNow(error = Dns::Name::AppendPointerLabel(aInfo.mHostNameOffset, *aInfo.mMessage));
+        ExitNow(error = Dns::Name::AppendPointerLabel(mHostNameOffset, *mMessage));
     }
 
-    aInfo.mHostNameOffset = aInfo.mMessage->GetLength();
-    SuccessOrExit(error = Dns::Name::AppendMultipleLabels(mHostInfo.GetName(), *aInfo.mMessage));
-    error = Dns::Name::AppendPointerLabel(aInfo.mDomainNameOffset, *aInfo.mMessage);
+    mHostNameOffset = mMessage->GetLength();
+    SuccessOrExit(error = Dns::Name::AppendMultipleLabels(mHostName, *mMessage));
+    error = Dns::Name::AppendPointerLabel(mDomainNameOffset, *mMessage);
 
 exit:
     return error;
 }
 
-Error Client::AppendUpdateLeaseOptRecord(MsgInfo &aInfo)
+Error Client::Update::AppendUpdateLeaseOptRecord(uint32_t aLease, uint32_t &aKeyLease)
```

Comment:
The fact that `AppendUpdateLeaseOptRecord()` can possibly modify its `aKeylease` input is somewhat unexpected and strange.

I understand why it was done. Before this change, this was a `Client` method that had access to `mKeylease`, so changing it in test mode was perhaps acceptable.

Now that this is a standalone helper method, the behavior feels odd. One idea is to handle this logic before calling the method. For instance, right before the call, we can check if `mUseShortLeaseOption` is active to set `mKeylease = mLease` and then pass that value to this method. What do you think?

Reply:
I have moved the code related to `mKeylease = mLease` to the end of the method `AppendServiceInstructions()`.